src/Projections/Reunion/Microsoft.UI.Dispatching.DispatcherQueueSynchronizationContext.DispatcherQueueProxyHandler.cs

```diff
@@ -0,0 +1,233 @@
+            /// </summary>
+            /// <returns>The initialized vtable pointer for <see cref="DispatcherQueueProxyHandler"/>.</returns>
+            /// <remarks>
+            /// The vtable itself is allocated with <see cref="RuntimeHelpers.AllocateTypeAssociatedMemory(Type, int)"/>,
+            /// which allocates memory in the high frequency heap associated with the input runtime type. This will be
+            /// automatically cleaned up when the type is unloaded, so there is no need to ever manually free this memory.
+            /// </remarks>
+            private static void** InitVtbl()
+            {
+                void** lpVtbl = (void**)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(DispatcherQueueProxyHandler), sizeof(void*) * 4);
+
+                lpVtbl[0] = (delegate* unmanaged<DispatcherQueueProxyHandler*, Guid*, void**, int>)&Impl.QueryInterface;
+                lpVtbl[1] = (delegate* unmanaged<DispatcherQueueProxyHandler*, uint>)&Impl.AddRef;
+                lpVtbl[2] = (delegate* unmanaged<DispatcherQueueProxyHandler*, uint>)&Impl.Release;
+                lpVtbl[3] = (delegate* unmanaged<DispatcherQueueProxyHandler*, int>)&Impl.Invoke;
+
+                return lpVtbl;
+            }
+
+            /// <summary>
+            /// The vtable pointer for the current instance.
+            /// </summary>
+            private void** lpVtbl;
+
+            /// <summary>
+            /// The <see cref="GCHandle"/> to the captured <see cref="SendOrPostCallback"/>.
+            /// </summary>
+            private GCHandle callbackHandle;
+
+            /// <summary>
+            /// The <see cref="GCHandle"/> to the captured state (if present, or a <see langword="null"/> handle otherwise).
+            /// </summary>
+            private GCHandle stateHandle;
+
+            /// <summary>
+            /// The current reference count for the object (from <c>IUnknown</c>).
+            /// </summary>
+            private volatile uint referenceCount;
+
+            /// <summary>
+            /// Creates a new <see cref="DispatcherQueueProxyHandler"/> instance for the input callback and state.
+            /// </summary>
+            /// <param name="handler">The input <see cref="SendOrPostCallback"/> callback to enqueue.</param>
+            /// <param name="state">The input state to capture and pass to the callback.</param>
+            /// <returns>A pointer to the newly initialized <see cref="DispatcherQueueProxyHandler"/> instance.</returns>
+            public static DispatcherQueueProxyHandler* Create(SendOrPostCallback handler, object? state)
+            {
+#if NET6_0
+                DispatcherQueueProxyHandler* @this = (DispatcherQueueProxyHandler*)NativeMemory.Alloc((nuint)sizeof(DispatcherQueueProxyHandler));
+#else
+                DispatcherQueueProxyHandler* @this = (DispatcherQueueProxyHandler*)Marshal.AllocHGlobal(sizeof(DispatcherQueueProxyHandler));
```

Comment:
What's the advantage of making a distinction here? Do we need two possible implementations?

Reply:
`Marshal.AllocHGlobal` is particularly slow and generally just not recommended for use anymore ([see here](https://github.com/dotnet/runtime/issues/33244#issuecomment-828619859)). `NativeMemory` has been introduced in .NET 6 and it's just a lightweigt set of wrappers for the relative C APIs. Once .NET 5 goes out of support (which is like in less than a month anyway) and WASDK drops it, we can just remove all of those extra paths anyway and just leave the ones for .NS2.0 (the default one) and the .NET 6 one. Does that make sense? 🙂